Actuators

coletaylor788 opened the pull request:
Adding controller node that outputs a CarPose and a CarState
Added files from buzzmobile-old for the arduino
Added CarState message

irapha committed:
copy over actuator code from buzzmobile-old

irapha committed:
add Horn to CarPose, and CarState message

coletaylor788 committed:
Adding files for controller

coletaylor788 committed:
added controller node to cmakelists

coletaylor788 committed:
Merge branch 'steering' into actuators

irapha committed:
actually add CarState msg

irapha committed:
Merge branch 'actuators' of github.com:gtagency/buzzmobile into actuators

irapha committed:
make Bool -> bool

coletaylor788 committed:
Adding instructions for PS4 controller setup

coletaylor788 committed:
Changing the controller node to work with new buzzmobile. Changing horn and main car application to use CarPose.msg

coletaylor788 committed:
Merge branch 'actuators' of https://github.com/gtagency/buzzmobile into actuators

coletaylor788 committed:
Removing files named differently

coletaylor788 committed:
Moving arduino files into folder per requirement. Adding state functionality to controller node.

irapha commented on buzzmobile/drive/horn/src/horn_node.cpp:
```diff
@@ -0,0 +1,42 @@
+#include <ros/ros.h>
+#include <std_msgs/Bool.h>
+#include <buzzmobile/CarPose.h>
```
This doesn't seem to compile:
https://travis-ci.org/gtagency/buzzmobile/builds/168462865

Maybe we need to `<depend>` on the manifest?

irapha commented on README.md:
````diff
@@ -48,6 +48,13 @@ sudo usermod -aG dialout <YOUR USERNAME>
+sudo pip install ds4drv
+sudo apt-get install ros-indigo-joy
+```
+
````
Can you add instructions for how the controller changes CarPose? Probs in README (in a new section, called `Manual Driving`, just like we have `Environment` and `Running`). Also mention how to change manual to auto modes

irapha commented on README.md:
````diff
@@ -48,6 +48,13 @@ sudo usermod -aG dialout <YOUR USERNAME>
+
+```bash
+sudo pip install ds4drv
+sudo apt-get install ros-indigo-joy
````
Can you add this `ros-indigo-joy` dependency to the `buzzmobile/package.xml`? Then I don't think we need to have it here because we can just install all deps declared there. All other deps like lidar and gps are there already.

Also, if `ds4drv` is a thing that's available as a rosdep (we'd have to look it up), then put it there too. Then we don't need this addition to the `README`.

coletaylor788 committed:
Updating travis.yaml to use catkin_make install

coletaylor788 committed:
Reverting change to catkin_make

coletaylor788 committed:
Adding dependencies to CMakeLists files

coletaylor788 committed:
Merge branch 'actuators' of https://github.com/gtagency/buzzmobile into actuators

coletaylor788 commented:
Whenever you create a ros msg that gets used by a cpp node, you need to add dependency to the CMakeLists.txt file for each Node that uses a custom message so it waits until after the messages are generated to compile. Otherwise catkin_make tries to compile the node .cpp file before the custom message .h file has been created. catkin_make aggresively attempts to parallelize.

add_dependencies(your_node ${catkin_EXPORTED_TARGETS})

http://answers.ros.org/question/62092/catkin_make-started-failing-to-generate-h-from-msg/

coletaylor788 committed:
Adding joy as a rosdep so it is automatically installed

coletaylor788 committed:
Update .travis.yml

Adding line to install ros dependencies

coletaylor788 committed:
Update .travis.yml

Moving rosdep install to after buzzmobile is set up.

coletaylor788 committed:
Trying to install dependencies after making

coletaylor788 committed:
Sourcing script to hopefully recognize buzzmobile

Pretty sure it needs to be after catkin_make so the devel directory exists.

coletaylor788 commented:
Added joy as a ros dependency in package.xml and updated .travis.yaml to install the ros dependencies. Note: I needed to source the script in catkin_ws/devel before installing the ros dependencies so it would recognize buzzmobile. Since I don't think devel exists until after catkin_make, I put this after the make command. (Not sure if there is a better alternative)

coletaylor788 committed:
Adding controller instructions

Also removed joy installation as it was unecessary after implementing as rosdep

irapha commented on .travis.yml:
```diff
@@ -28,6 +28,8 @@ install:
 script: # All commands must exit with code 0 on success. Anything else is considered failure.
   - catkin_make -j2
+  - source ~/catkin_ws/devel/setup.bash # Allow installing dependencies (recognizing buzzmobile)
+  - rosdep install buzzmobile # Install dependencies for buzzmobile
```
should we be installing dependencies before making? I'd assume we install deps in `before_script`, no?

irapha replied:
As discussed offline, you can only install deps after you source the `setup.bash`, which is in the `devel/`, which is generated from `catkin_make`. So the way this is done seems correct.

At one point, when we have rostests ( see #20 ) we will move these three script commands to `before_script` and have the test command be the `script`.

coletaylor788 committed:
Editing main.cpp to connect to arduino and ino file to skip if statement check temporarily

coletaylor788 committed:
Merge branch 'actuators' of https://github.com/gtagency/buzzmobile into actuators

coletaylor788 committed:
Adding brake to car pose, adding brake to controller_node, changing max turning radius in controller_node

coletaylor788 committed:
Moving controller to plan/

coletaylor788 committed:
Adding mux and changing steering publisher to auto_car_pose

joshuamorton commented on buzzmobile/sense/controller/src/controller_node.cpp:
```diff
@@ -10,35 +10,34 @@
-//NOTE: this is also published by the autonomous brake...this is ok.
-ros::Publisher brake_pub;
+
+const float MAX_TURNING_ANGLE = M_PI / 6; // 30 degrees in radians
```
Factor this out into constants.yml?

irapha replied:
See discussion on #16

joshuamorton commented on buzzmobile/sense/controller/src/controller_node.cpp:
```diff
@@ -10,35 +10,34 @@
 
 bool obstacleFlag = false;
 
+// Keep track of the values to continously publish
```
I think we actually shouldn't need that, ROS will cache old values for us, but not a major issue.

irapha replied:
Hm. Idk what happens if we go with either case. If we don't continuously publish a message to the motor, won't the motor get the message and then stop doing what it was doing because it didn't get a message right after telling it to continue at it?

If this doesn't happen, then I agree this thing isn't needed.

irapha committed:
a few fixes and factorings in car_pose_mux

irapha committed:
typo

irapha committed:
fix conflicts

irapha committed:
Merge branch 'master' into actuators

irapha committed:
revert travis changes

irapha committed:
Fix joy subscriber in preparation for launch files

coletaylor788 committed:
Adding directions to enabled car to README

coletaylor788 committed:
Fixing formatting of README

coletaylor788 committed:
Fixing car_pose_mux and merging horn into main.cpp

coletaylor788 committed:
Merge branch 'actuators' of https://github.com/gtagency/buzzmobile into actuators

coletaylor788 committed:
Fixing car pose to include START. Changing controller to start in START state

irapha commented:
Can you merge master into actuators and fix conflicts?

irapha commented on README.md:
```diff
@@ -135,3 +141,25 @@ To see info about the recorded data, do `rosbag info filename.bag`
+2. Flip the switch inside to turn the car on
+3. Ensure all e-stops are disabled. (Red buttons on front and back in out position and enabled via remote)
+4. Press the green button to start the motors (Car is now live)
+5. See Controller for details on how to drive the car once enabled.
```
Mention something about car starting in START mode, and that you need to switch to either MANUAL or AUTO mode in the controller?

irapha commented on README.md:
```diff
@@ -135,3 +141,25 @@ To see info about the recorded data, do `rosbag info filename.bag`
+4. Press the green button to start the motors (Car is now live)
+5. See Controller for details on how to drive the car once enabled.
+
+Controller
```
Change title to "Manual Mode Controls"?

irapha commented on README.md:
```diff
@@ -135,3 +141,25 @@ To see info about the recorded data, do `rosbag info filename.bag`
+Controller
+----------
+
+The controller_node node outputs a CarPose message and a CarState message determined by input from a PS4 controller. To control these messages and operate the car manually using the controller, use the following controls:
```
controller_node => `controller_node`

irapha commented on buzzmobile/CMakeLists.txt:
```diff
@@ -31,4 +32,8 @@ include_directories(
+
 add_subdirectory(process/projector)
+add_subdirectory(plan/controller)
+add_subdirectory(drive/horn)
```
Didn't we remove `horn` node?

irapha commented on buzzmobile/drive/arduino/src/car_driver/car_driver.ino:
```diff
@@ -0,0 +1,158 @@
+   * This prevents the controllers ramping up while the
+   * vehicle is estopped, creating a dangerous situation.
+   */
+  //if(digitalRead(estop_pin) == HIGH) {
```
Don't merge to master with this commented out. We'll wait on @sastorer to fix the serial 6 connection and test this before we merge this PR.

irapha commented on buzzmobile/drive/car_interface/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+#rosbuild_add_library(${PROJECT_NAME} src/example.cpp)
+#target_link_libraries(${PROJECT_NAME} another_library)
+#rosbuild_add_boost_directories()
+#rosbuild_link_boost(${PROJECT_NAME} thread)
```
Do we need all this commented out stuff?

irapha commented on buzzmobile/drive/car_interface/src/main.cpp:
```diff
@@ -0,0 +1,88 @@
+  msg.header.seq = odom_sequence++;
+  msg.header.stamp = ros::Time::now();
+  encoder_pub.publish(msg);
+}*/
```
Just a note that I'm okay with the commented out stuff on this file since there's an open issue for adding this in the future #41 #66

irapha commented on buzzmobile/drive/horn/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+#rosbuild_gensrv()
+
+#common commands for building c++ executables and libraries
+#rosbuild_add_library(${PROJECT_NAME} src/example.cpp)
```
Do we need this commented out stuff?

irapha commented on buzzmobile/drive/horn/src/horn_node.cpp:
```diff
@@ -0,0 +1,42 @@
+  ros::Duration(0.5).sleep(); // sleep for half a second
+  car_horn_pub.publish(off);
+//  std::cout << "Exiting honkLong" << std::endl;
+}*/
```
Also fine with these commented out things because we have open issues for handling these in the future #41 #66

irapha commented on buzzmobile/plan/car_pose_mux/car_pose_mux.py:
```diff
@@ -0,0 +1,47 @@
+def set_car_state(car_state):
+    state['curr_car_state'] = car_state.state
+    rospy.loginfo("Car in " + (
+        "AUTO" if car_state.state == CarState.AUTO else "MANUAL") + " mode.")
```
This will only print either MANUAL or AUTO. It should also print START.

irapha commented on buzzmobile/plan/controller/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+
+#uncomment if you have defined services
+#rosbuild_gensrv()
+
```
so many comments y pls

irapha commented on buzzmobile/plan/controller/src/controller_node.cpp:
```diff
@@ -0,0 +1,193 @@
+//		honkHorn();
+//	} //TODO: for some reason this will honk repeatedly (repeated obstacles)...may need to debug this
+//	
+//	obstacleFlag = flag->data;
```
We might want another set of eyes in this file. @joshuamorton @dacohen Do you guys know what this commented out stuff does? They were in the controller in `buzzmobile-old`.

joshuamorton commented:
One of the nice things Ros does is that messages are stored in a queue, and
that queue will never be empty once it is filled, that's why ros.spinonce
works.

On Fri, Oct 21, 2016, 11:31 Raphael Gontijo Lopes notifications@github.com
wrote:

> ## _@iRapha_ commented on this pull request.
> 
> In buzzmobile/sense/controller/src/controller_node.cpp
> https://github.com/gtagency/buzzmobile/pull/31:
> 
> >  #define horn_button buttons[1]
> >  #define velocity_trigger axes[4]
> >  #define toggle_auto_button buttons[12]
> 
>  ros::Publisher motion_pub;
>  ros::Publisher state_pub;
> -//NOTE: this is also published by the autonomous brake...this is ok.
> -ros::Publisher brake_pub;
> +
> +const float MAX_TURNING_ANGLE = M_PI / 6; // 30 degrees in radians
> 
> See discussion on #16 https://github.com/gtagency/buzzmobile/issues/16
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/gtagency/buzzmobile/pull/31, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AEz4ACqAsWnHuOoPlmm4lE3H7OHwuSjBks5q2NrNgaJpZM4KZK72
> .

coletaylor788 committed:
Fixing merge conflicts

coletaylor788 committed:
Updating README for more details

coletaylor788 committed:
Removin horn node, deleting comments, adding START to mux output

coletaylor788 committed:
Merge branch 'actuators' of https://github.com/gtagency/buzzmobile into actuators

coletaylor788 committed:
Making slight changes to README

coletaylor788 committed:
Changing controller.py to controller.cpp

coletaylor788 committed:
Removing architecture diagram to refresh it

coletaylor788 committed:
Re-adding architecture image to refresh it

irapha commented on buzzmobile/plan/car_pose_mux/car_pose_mux.py:
```diff
@@ -33,8 +33,14 @@ def set_auto_car_pose(car_pose):
+        mode = "MANUAL"
+    else:
+        mode = "START"
+    rospy.loginfo("Car in " + mode + " mode.")
```
Maybe factor out this logging? This method got kinda unnecessarily big

coletaylor788 committed:
Factoring out car_state logging

coletaylor788 committed:
Merge branch 'actuators' of https://github.com/gtagency/buzzmobile into actuators

coletaylor788 committed:
Fixing style of factor out
